### ISSUE TITLE
Get master branch from github

### ISSUE THREAD

**Issue by gnzlbg** [ruslo/hunter#203]
Some packages like e.g. range-v3 don't have version. I would like to fetch the master branch from github with Hunter. Is this possible? From reading the docs I don't see any examples about how to do this. 

Right now I'm just doing the following myself:

```cmake
# Find the range-v3 include directory
# The following variables are set if range-v3 is found.
#  range-v3_FOUND        - True when the range-v3 include directory is found.
#  range-v3_INCLUDE_DIR  - The path to where the meta include files are.
# If range-v3 is not found, range-v3_FOUND is set to false.

find_package(PkgConfig)

if(NOT EXISTS "${range-v3_INCLUDE_DIR}")
  find_path(range-v3_INCLUDE_DIR
    NAMES range/v3/range.hpp 
    DOC "range-v3 library header files"
    )
endif()

if(EXISTS "${range-v3_INCLUDE_DIR}")
  include(FindPackageHandleStandardArgs)
  mark_as_advanced(range-v3_INCLUDE_DIR)
else()
  include(ExternalProject)
  ExternalProject_Add(range-v3
    GIT_REPOSITORY https://github.com/ericniebler/range-v3.git
    TIMEOUT 5
    CMAKE_ARGS -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER} -DCMAKE_CXX_FLAGS=${CMAKE_CXX_FLAGS}
    PREFIX "${CMAKE_CURRENT_BINARY_DIR}"
    CONFIGURE_COMMAND "" # Disable configure step
    BUILD_COMMAND "" # Disable build step
    INSTALL_COMMAND "" # Disable install step
    UPDATE_COMMAND "" # Disable update step: clones the project only once
    )
  
  # Specify include dir
  ExternalProject_Get_Property(range-v3 source_dir)
  set(range-v3_INCLUDE_DIR ${source_dir}/include)
endif()

if(EXISTS "${range-v3_INCLUDE_DIR}")
  set(range-v3_FOUND 1)
else()
  set(range-v3_FOUND 0)
endif()
```

**Comment by ruslo** [comment]
@gnzlbg Hi, thanks for report.

> Some packages like e.g. range-v3 don't have version

Hunter designed so that your root directory is fully reproducible (e.g. see [1](https://github.com/ruslo/hunter#what-is-it) and [2](https://github.com/ruslo/hunter#uninstall)). This is not consistent with the intention to just get source by git fetch from master branch. You have two choices:
* ask the upstream developer to make a release
* create a fork (see "Hunterized" [forks](https://github.com/hunter-packages)) and do release yourself

Notes:

1. Though this project use CMake as a build system I see no install target. It is a quite wrong widespread practice about header-only libraries since header-only library can be installed with `*Config.cmake` module which help users to find it using `find_package(... CONFIG)`, see [wiki](https://github.com/ruslo/hunter/wiki/example.find_package.config#cmake).

2. See planned feature (git SHA1/submodule part): https://github.com/ruslo/hunter/issues/56

**Comment by ruslo** [comment]
Not related directly but reminds me about some thoughts: https://github.com/ruslo/hunter/issues/204

**Comment by gnzlbg** [comment]
> Hunter designed so that your root directory is fully reproducible (e.g. see 1 and 2). 

Thanks for the links! Does this mean that I do not get a root directory per project but a single root directory for all my projects instead?

> Though this project use CMake as a build system I see no install target.

I don't want to install the dependency. In general, I want to have header only libraries in the build directory of my project.One could say that for header only libraries I would like to have "per-project" root directories. For the master and development branches I want to track upstream of my dependencies as closely as possible to fill issues right when something breaks (and not 6 months later when I try to upgrade to a release). The advantage of having them in a per project root is that I can just branch directly in my build directory, fix the library, commit to my repo and send a PR. I can also switch the FindXXXX.cmake file to point to my fork of the library while the PR is accepted. 

Other projects using their own "copies" of the library (either on some root directory or in their build directory) don't see any of this. 

When I want to make a release I typically just fix an upstream commit of the header only library in the FindXXXX.cmake files so that I can get a reproducible build for a release. But this is a rare situation (compared to development).

**Comment by ruslo** [comment]
> I do not get a root directory per project but a single root directory for all my projects instead?

Yes and no :) 

There is a single `HUNTER_ROOT` directory where lives other `<hunter-id>/<config-id>/<toolchain-id>` directories. I.e. if you build boost 1.58 for iOS 8.2 it will goes, say, to "${HUNTER_ROOT}/_Base/1eae623/0fa873a/**c39da39**" and any iOS 8.2 based projects can share this directory. And if you build boost 1.58 for Android it will goes, say, to "${HUNTER_ROOT}/_Base/1eae623/0fa873a/**c018e63**" and all Android based projects will reuse it.

See [customization levels](https://github.com/ruslo/hunter#features):
* [Hunter-ID](https://github.com/ruslo/hunter#hunter-id)
* [Config-ID](https://github.com/ruslo/hunter#config-id)
* [Toolchain-ID](https://github.com/ruslo/hunter#toolchain-id)

**Comment by ruslo** [comment]
> In general, I want to have header only libraries in the build directory of my project

I don't see the point of using ExternalProject_Add or package manager in this case. Just add a submodule and include directory by `include_directories("3rdParty/<package>/include")`.

**Comment by gnzlbg** [comment]
Git submodules are horrible, the only situation in which I would consider them is for creating "super projects" (like modular boost, or opencv), but not for any self contained project. Git subtree is a bit better, but with ExternalProject_Add people just don't have to worry about these issues at all. If I want to fix a git commit, a git tag, the git HEAD, or change to a different branch or fork of a project, I easily can do this by changing one line of code, without polluting my repo, and without users and other devs having to care about this at all.

**Comment by ruslo** [comment]
> Git submodules are horrible, the only situation in which I would consider them is for creating "super projects" (like modular boost, or opencv), but not for any self contained project

I agree in general that submodules have some flaw in design when used in package version management but for your particular case it's fit best. In abstract:
* you want to push updates to upstream - you must have a git repository with `range-v3` history somewhere
* you want to be able to share your development code with others. I.e. you create branch `feature-1` with some updates in `range-v3` and push it. Other developer pull your changes, hence expect to have the **exact** version of `range-v3` - i.e. you must store SHA1 of `range-v3` somewhere. Note that your currect CMake code with ExtrenralProject_Add is not stable. The result library will change from time to time because it depends just on master branch instead of SHA1 or tag.

> Git subtree is a bit better

Haven't used it in practice. Does it solves next situation:
* A depends on X v1.0
* B depends on X v2.0
* C depends on A and B
what version of X will use C? How this conflict solved?

>  If I want to fix a git commit, a git tag, the git HEAD, or change to a different branch or fork of a project, I easily can do this by changing one line of code

Yes, manually. With submodule you can just do 'git commit' in child project, that's all.

> I have yet to find a draw-back for ExternalProject_Add for header-only libraries

ExternalProject_Add is good basic block but it's not extendable for something more. Superbuilds are quite hard to maintain and reuse (guess you have two projects with superbuilds, how you merge them into one?). Also ExternalProject_Add doesn't track dependencies. Once project installed it creates DONE stamps. If dependency change you have to remove this stamps manually somehow to trigger rebuild procedure.

Also if you will extend your example to the "real" library (I mean with some libraries compiled from sources) you will find out that on configure stage **there is no** real files and you have to "predict" their location manually. Even in currect example value of `range-v3_FOUND` is `0` after `cmake -H. -B_builds`.

**Comment by ruslo** [comment]
> Does this mean that I do not get a root directory per project but a single root directory for all my projects instead?

Also you can set value of `HUNTER_ROOT` CMake variable to any directory. So in general you can set it per project to the `CMAKE_CURRENT_LIST_DIR/3rdParty` (though it's not optimal since root is not shared).

**Comment by gnzlbg** [comment]
The problem with git submodules is that it is a good workflow for the 1% of the situations that almost never happen, and annoying/bad for anything else. ExternalProject_Add is perfect for 99% of what happens every day, and pretty good at the 1% that never happens. 

> Note that your currect CMake code with ExternalProject_Add is not stable.

But that is the whole point. Stability is only important when you are doing a release. How often do you release vs How often do you commit during development? 

When I branch out into a release branch, which happens max twice a year, I can add a git commit or a git tag to ExternalProject_Add to get stability. The other 100% of the time I just want to track the master branch of my dependencies to prevent them from screwing my project up. ExternalProject_Add just works without issues.

But git submodules:
- forces users (and external package managers) to know that I am using git submodules to even perform a correct checkout of my library.
- is extremely noisy. Your repository gets filled with 10s of commits per day that just say "updated submodule X to master". 

This is a pain to use every single day of the year except those two days I do a release. ExternalProject_Add is a breeze to use every single day of the year. 

> A depends on X v1.0
> B depends on X v2.0
> C depends on A and B what version of X will use C? How this conflict solved?

If the ABIs are incompatible _how do you_ solve this conflict? A and B cannot interoperate. 

Yes, this is a tough problem, but one of the advantages of tracking the master branch of your libraries is that you can force them to use the same version of their dependencies. That is, when you want to do a release, all your libraries depend on X v3.0 and this problem never appears. Another advantage is that modern versions use inline namespaces, which just removes this problem completely.

> you want to be able to share your development code with others. I.e. you create branch feature-1 with some updates in range-v3 and push it. Other developer pull your changes, hence expect to have the exact version of range-v3 - i.e. you must store SHA1 of range-v3 somewhere.

Again, how often do you find a bug in a dependency and need to commit to your master or development branch code with a bug fix? To me it happens 1% of the time I spend developing. ExternalProject_Add gives me multiple alternatives: I can change the repo to one with the fix or to one that used to work. I really don't see how these alternatives are worse that what git submodules gives me. Anyone checking out the repo will get a FindXXX.cmake file that gets them a dependency that works. Those checking a release version don't have this problem since the FindXXX.cmake files specify a commit that just works.

>  to the "real" library (I mean with some libraries compiled from sources)

Are googletest and boost real libraries? I used them just fine with this approach.

**Comment by ruslo** [comment]
> Stability is only important when you are doing a release

I see and understand your point, but will all respect I don't agree with that. There are tons of factors which make life of developer full of pain and surprises and I don't want to add one more random variable to that equation. I can do 'git submodule update' from time to time and do "3rd party updated" commit, see no problem here and I can't said it's too annoying or noisy. I guess it's just a matter of experience. Once you'll trap in situation when "something suddenly brokes and I can't find what is broken exactly but need to make some important fix immediately" you will understand what I am talking about :)

> If the ABIs are incompatible how do you solve this conflict?

Who said they are not compatible? Guess A and B both can work with X v1.0 and v2.0. The point is that git submodule defines dependency locally. Version of X hardcoded in A. Same with B.

> Yes, this is a tough problem

I'm not sure what exactly you mean but situation that I've described is not a problem for Hunter (to be precise either it will work correctly or will report a problem and will not work at all).

> but one of the advantages of tracking the master branch of your libraries is that you can force them to use the same version of their dependencies

Note that same happens with your approach. Your project A depend on range-v3 and do download master branch from git. There is project B which use same approach and download range-v3 from git, but set version tag to v1.0. Now somebody want to use both A and B...

> Another advantage is that modern versions use inline namespaces, which just removes this problem completely

different inline namespace for each commit?

> Are googletest and boost real libraries? I used them just fine with this approach

GoogleTest is one file library. Can you show your ExternalProject_Add for Boost? What libraries does it support? Does it support Linux/Windows? Does it support iOS/Android? You have to "predict" all the libraries and all components, I guess it's a huge pain.

Again even your current CMake code returns `range-v3_FOUND = 0` do you think it's normal?

---

So to summarize:
The closest solution I can offer is this enhancement: https://github.com/ruslo/hunter/issues/56. I guess I can even add branch name (instead of raw SHA1 or submodule) but it will make sense only with "dynamic hunter-id"  feature: https://github.com/ruslo/hunter/issues/204. Also you can have all the 3rd party libraries locally by setting `HUNTER_ROOT` to some location like `CMAKE_CURRENT_LIST_DIR/3rdParty`. Feel free to subscribe to this bugs to see when they will be solved (or you can even implement them).

Cheers, Ruslan

**Comment by gnzlbg** [comment]
> I see and understand your point, but will all respect I don't agree with
that.

It is ok to disagree, with understanding everyone grows.


> X v1.0 and v2.0.


Sorry, I assumed this was semantic versioning. Under your assumption the
latest version always works for both.

> different inline namespace for each commit?

No, only on an API/ABI breaking change. Inline namespaces allow you to use
multiple ABI incompatible versions of the same library at the same time by
just linking against the latest version of the library. It even allows the
different versions to communicate between them. (E.g. pass a
std::v2::string to a function taking an ABI incompatible std::v3::string).

> Can you show your ExternalProject_Add for Boost? What libraries does it
support? Does it support Linux/Windows? Does it support iOS/Android? You
have to "predict" all the libraries and all components, I guess it's a huge
pain.

It supports all the header only libraries. IIRC it just fetches the boost
superproject from github, and for every directory adds the "include"
directory to a cmake variable.

> Again even your current CMake code returns range-v3_FOUND = 0 do you
think it's normal?

Thanks for the fix! It is working correctly now!

> Feel free to subscribe to this bugs to see when they will be solved (or
you can even implement them).

Thanks, subscribed!

**Comment by ruslo** [comment]
> Under your assumption the latest version always works for both

It's not the point. The main issue is that both libraries should use the same version at the same time otherwise it violates One Definition Rule. Any methods of defining version of library locally (git submodules, git tree or manual downloading to local directory) violates this rule when version of dependency defined in several places.

> No, only on an API/ABI breaking change

Even if API/ABI doesn't change I think it's not a good idea to have 2 or more versions of one library in one project.

> It supports all the header only libraries

Then you'll be **very** surprised when times comes to add such libraries as system, filesystem, log, regex, ...

> Thanks for the fix! It is working correctly now!

What fix? Code from you first comment still result `range-v3_FOUND=0`

**Comment by gnzlbg** [comment]
I'm closing this since I just wanted to ask if Hunter can support my use cases and it seems it follows a very different philosophy which is just fine. It is a philosophy I understand, but while I can live with header only dependencies, one could say I am more on the "no dependencies" side of the spectrum. 

> The main issue is that both libraries should use the same version at the same time otherwise it violates One Definition Rule. 

In C++03 one just use the latest version of both (how is this a hard problem?), and in C++11 one can use both at the same time without violating the One Definition Rule (as long as they version their namespaces). To me the only interesting problem is how does one use two ABI incompatible versions of the same library at the same time, and how does one pass types of one to the other. The only solution I know of is `inline namespaces`, which is pretty easy to use, I really don't know how to solve this at the build system level (if it is possible at all, it will be hard and painful). 

> Then you'll be very surprised when times comes to add such libraries as system, filesystem, log, regex, ...

Surprised? I explicitly decided to use the header only libraries to not be able to use these. First if I ever need two packages that depend on two ABI incompatible versions of Boost there is nothing a build system can do. To me it is an objective to don't let it come to this point. Second, there are header only alternatives to some of these libraries.

> Even if API/ABI doesn't change I think it's not a good idea to have 2 or more versions of one library in one project.

Agreed.

> What fix? Code from you first comment still result range-v3_FOUND=0

I haven't committed it yet, I just set it to found after calling external project add.

**Comment by ruslo** [comment]
> In C++03 one just use the latest version of both (how is this a hard problem?)

Because with the approach you're currently using the version is hardcoded into the project. Say you have header-only library `foo` with version `v1.0.0`:

```cpp
inline int foo() {
  return 3;
}
```

and version `v1.0.1`:

```cpp
inline int foo() {
  return 4;
}
```

project `A` use version `v1.0.0` which is **hardcoded** in your ExternalProject_Add file (we are talking about stable version with releases) and holds only one function in a.cpp file (`A` is a library):

```cpp
int a() {
  std::cout << "a: foo returns " << foo() << std::endl; // -> 3
}
```

same with the project `B`, which depends on `v1.0.1`:

```cpp
int b() {
  std::cout << "b: foo returns " << foo() << std::endl; // -> 4
}
```

They both works fine. But if you want to add both of them to the project `C` - you will have a problems:

```cpp
int main() {
  a();
  b();
}
```

result will be:

```
a: foo returns 3
b: foo returns 4
```

This is called violation of One Definition Rule. Note that `foo` version doesn't break API.

**Comment by ruslo** [comment]
> and in C++11 one can use both at the same time without violating the One Definition Rule (as long as they version their namespaces)

Add namespace with some version to the previous example and nothing will change.

> I explicitly decided to not allow the binary libraries

Binaries exist for the reason. You decide to not use binaries? Okay - goodbye OpenCV, Boost (system, filesystem, regex, log, ...), Qt, OpenSSL, ...

> First, if I ever need two packages that depend on two ABI incompatible versions of Boost there is nothing a build system can do

If it's already built - yes. However Hunter builds incompatible versions in separate directories (see [toolchain-id](https://github.com/ruslo/hunter#toolchain-id)). So if you want to have two versions of Boost: one for 'clang -std=c++11 -stdlib=libc++' and other 'gcc' - it's not a problem.

> but it will exist when it is need it

If just an indicator of the problem. You set it to `TRUE`, ok. It's called "workaround", in this particular case it's called hack :) For example this code will not work too:

```cmake
if(range_v3_FOUND)
  file(READ "${range-v3_INCLUDE_DIR}/range-v3/version.h" RANGE_V3_VERSION)
  message("Using range-v3 version: ${RANGE_V3_VERSION}") # nothing, since file not exists yet
endif()
```

**Comment by gnzlbg** [comment]
> Add namespace with some version to the previous example and nothing will change.

Let's give that a try

```c++
namespace foo {
namespace v1 { inline int foo() { return 3; } }
namespace v2 { inline int foo() { return 4; } }
inline // maybe allow configuring this with a macro
namespace v3 { inline int foo() { return 5; } }
}

namespace A { 
using namespace foo::v1;  // maybe allow configuring this with a macro 
int a() {  std::cout << "a: foo returns " << foo() << std::endl; } 
} 
namespace B { 
using namespace foo::v2; 
int b() {  std::cout << "b: foo returns " << foo() << std::endl; } 
} 
```

What did change? For a starter, the one definition rule is now satisfied, so this is now a valid program (before it was ill-formed because it broke ODR). In particular: Library A can use v1::foo() and Library B can use v2::foo(), all within the same binary. 

If my program requires `foo` it will get the latest version automatically (which is `v3::foo`) or I can fix the version I want to use explicitly.  

> Note that foo version doesn't break API.

It doesn't break the ABI (type signature) but if foo() is defined as "always returns 3" and now it is defined as "always return 5" that certainly is an API break. If it was defined as "returns some number between 1 and 10" then that would not be an API break.

> Binaries exist for the reason. You decide to not use binaries? Okay - goodbye OpenCV, Boost (system, filesystem, regex, log, ...), Qt, OpenSSL, ...

I said I do not use the binary versions of these Boost libraries. In particular, because better alternatives exist and these are header only. When a library I use or a project I work on needs a binary library, I haven't had a problem requiring these libraries via CMake yet. When the library is not found, the number of alternative ways to proceed is just too large.

> If just an indicator of the problem. You set it to TRUE, ok. It's called "workaround", in this particular case it's called hack :) For example this code will not work 

Agreed, but not all problems are worth solving. If a dependency of mine wants to do that, it is as easy as having the library installed (that way it won't be fetched from the internet by my project).

**Comment by ruslo** [comment]
> What did change?

I have a patch fix. It's just one commit that fix some minor problem. So I should change `v1` to `v2` for this? We are returning to what I've said earlier: `different inline namespace for each commit?` Of cource if I will protect my library on each commit with new tag we will not violate ODR. In the case above it should be `namespace v1_0_0` and `namespace v1_0_1`.

> If it was defined as "returns some number between 1 and 10" then that would not be an API break

How does it make a difference? Okay the API said "foo returns some number between 1 and 10". So it's not an API break, yes? But we still have an ODR violation. Don't undertand this argument.

> I can fix the version I want to use explicitly

Take a look at my example again. You have to set version of this library for **every** project that use it. **Manually**. I.e. if you have `A`, `B` and `C` depends on `foo` you must be sure that all of them use tag `v3` in their ExternalProject_Add commands before combining them together.

> When a library I use or a project I work on needs a binary library, I haven't had a problem requiring these libraries via CMake yet. When the library is not found, the number of alternative ways to proceed is just too large
...
> Agreed, but not all problems are worth solving. If a dependency of mine wants to do that, it is as easy as having the library installed (that way it won't be fetched from the internet by my project)

To summarize: you said "ExternalProject_Add" fit best for you and you have no problems with it. I said you will have problems with binaries and the fact that even header-only library does not exist physically on configure step. Your reply: "I will solve binary problems somehow in alternative way" and "users will install my library easily (somehow, again)". I'm a little bit confused... :)

**Comment by gnzlbg** [comment]
On Wednesday, July 15, 2015, Ruslan Baratov <notifications@github.com>
wrote:

> What did change?
>
> I have a patch fix. It's just one commit that fix some minor problem. So I
> should change v1 to v2 for this?
>

No, for the _third time_ only on an API or ABI breaking  change.


> We are returning to what I've said earlier: different inline namespace
> for each commit?
>
>
No, read above.


> Of cource if I will protect my library on each commit with new tag we will
> not violate ODR. In the case above it should be namespace v1_0_0 and namespace
> v1_0_1.
>
> If it was defined as "returns some number between 1 and 10" then that
> would not be an API break
>
> How does it make a difference?
>
>
It does not break the API or the ABI.



> Okay the API said "foo returns some number between 1 and 10". So it's not
> an API break, yes? But we still have an ODR violation.
>
> No, there is no ODR violation.


> Don't undertand this argument.
>
> I can fix the version I want to use explicitly
>
> Take a look at my example again. You have to set version of this library
> for *every* project that use it.
>
> No, each project fix it for itself.
You don't have to do anything.


> *Manually*. I.e. if you have A, B and C depends on foo you must be sure
> that all of them use tag v3 in their ExternalProject_Add commands before
> combining them together.
>
No, they can use different versions, they decide which version to use, and
there are no ODR violations. You don't have to do anything anywhere.

 When a library I use or a project I work on needs a binary library, I
> haven't had a problem requiring these libraries via CMake yet. When the
> library is not found, the number of alternative ways to proceed is just too
> large
> ...
> Agreed, but not all problems are worth solving. If a dependency of mine
> wants to do that, it is as easy as having the library installed (that way
> it won't be fetched from the internet by my project)
>
> To summarize: you said "ExternalProject_Add" fit best for you and you have
> no problems with it. I said you will have problems with binaries and the
> fact that even header-only library does not exist physically on configure
> step. Your reply: "I will solve binary problems somehow in alternative way"
>
> No, I reply my users can solve this problems in the best way they see fit.
This is better than solving the problems for them in the wrong way.
Automatically downloading a binary library and installing it is the wrong
choice in a lot of situations.

> and "users will install my library easily (somehow, again)". I'm a little
> bit confused... :)
>
 I can see that.

>

**Comment by ruslo** [comment]
> No, for the _third time_ only on an API or ABI breaking change

I'll try it another way. Assume we have header-only library `foo` with function `int foo()`.

Version `v1.0.0`:

```cpp
// file foo.hpp
inline int foo() {
  return 3;
}
```

Now we are applying patch and release version `v1.0.1`:

```cpp
// file foo.hpp
inline int foo() {
  return 4;
}
```

Do we breaking API/ABI compatibility?

**Comment by gnzlbg** [comment]
That never breaks the ABI.

Wether that breaks the API or not depends on the contract of the function.

Can the users rely on a particular integer or range of integers being
returned or not?

You don't specify it.  If they cannot, that does not break the API. If they
can, that might or might not break the API depending on the particular
contract.

**Comment by ruslo** [comment]
> Wether that breaks the API or not depends on the contract of the function

I.e. it depends on such stuff like documentation? Interesting... Okay this function can return any integer number. So there is no need to add a namespace on this patch, right?

Also take a look at the quote from [kde docs](https://techbase.kde.org/Policies/Binary_Compatibility_Issues_With_C%2B%2B#The_Do.27s_and_Don.27ts):

```
* change an inline function or make an inline function non-inline if it is safe that
programs linked with the prior version of the library call the old implementation.

This is tricky and might be dangerous. Think twice before doing it
```
